### DISCUSSION TITLE
Issue with architecture tests: `dependsOnNothing` always pass

### DISCUSSION THREAD

**Discussion by jmcazaux** [LemonAppDev/konsist#1423] [Report a Problem] [answered]
### Tell us what version of Konsist you use.

0.16.1

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

Hi there,
I am experimenting with Konsist for a large multi module project.
The first attempt with declaration assertions went well, really helpful at the first glance 👍!

Unfortunately, I was not so lucky with architecture tests.

In a nutshell, whatever I test, the test pass.
I have recreated the problem in a minimal project [here](https://github.com/jmcazaux/konsist-test).
The [Konsist tests](https://github.com/jmcazaux/konsist-test/blob/main/architecture/src/test/kotlin/ArchitectureTests.kt) are in the `architecture` submodule. 

I have printed the scope, it is OK.
I have as well tried with a `Main` class (instead of the function), to no avail.

Thanks in advance for your help.

### Actual behaviour

When the `app` subproject (& layer) does include `lib`, the below test passes (when it should fail).
```
scope.assertArchitecture {
    val appLayer = Layer("app", "com.ironbird.app..")
    appLayer.dependsOnNothing()
}
```

When the `lib` subproject (&layer) does **not** include `app`, the below test passes (when it should fail, or `dependsOn` is a _can_ and not a _must_).
```
scope.assertArchitecture {
    val appLayer = Layer("app", "com.ironbird.app..")
    val libLayer = Layer("lib", "com.ironbird.lib..")
    libLayer.dependsOn(appLayer)
}
```

### Expected behaviour

Both tests above should fail.

**Comment by jmcazaux** [comment]
In case it helps, I have improved the repro-case a bit.

It is a very simple setup with an `app` module containing the `app` package and a `lib` module containing the `lib` package.
`app` imports `lib`.
`lib` imports nothing.

When I test that:
- `app.doesNotDependOn(lib)`, it **fails**: that is expected.
- `app.dependsOnNothing()`, it passes: that should not be the case IMHO (as the above proves `app` depends on `lib`)
- `lib.dependsOn(app)`, it passes: that should not be the case, unless the `DependencyRules.Layer.dependsOn` means _can depend on_ and not _must depend on_.

Hope this helps.

**Comment by igorwojda** [comment]
This is a valid bug. 
Architecture checks (`assertArchitecture`) require some polishing, rest of the Konsist API is much more polished.

We will take a closer look, but it will take some time as we are wrapping up next release. Perhaps you want to try to debug this and open PR (good starting point would be looking at `Architecture1Test` class and for example trying to reproduce this inside Konsist project. Also `Konsist.scopeFromDirectory` / `Konsist.scopeFromExternalDirectory` may be helpful for reproducing this. Another starting point may be removing Konsist dependency from you project and copy code from `libs` module to facilitate debugging)?

BTW You can use `Konsist.scopeFromProduction()` instead of `Konsist.scopeFromProject() - Konsist.scopeFromTest()`

**Reply by jmcazaux** [reply]
Many thanks for this @igorwojda I'll try to have a look.

There is no other option than `assertArchitecture` to run an architecture test, right?

**Reply by igorwojda** [reply]
Correct

**Comment by igorwojda** [comment, marked as answer]
@jmcazaux I have [heavily refactored](https://github.com/LemonAppDev/konsist/pull/1505) code around `assertArchitecture`

I have opened PR to your project with `0.17.0-SNAPSHOT` that is fixing reported issues https://github.com/jmcazaux/konsist-test/pull/1.  Please give it a try.

BTW your sample was extremely helpful in fixing this issue.

**Reply by jmcazaux** [reply]
Thanks for this @igorwojda, I'll have a look.
Unfortunately, time is sparse these days ;).
